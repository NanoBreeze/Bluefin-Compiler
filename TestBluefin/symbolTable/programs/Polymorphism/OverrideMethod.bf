/*
This program tests that the 'override' keyword is used correctly. Specifically, it tests that
- override can occur only in methods, not functions
- override applies only if the method truly does override the base class' method with same name, return type, and param types
*/

struct Base {
	void f(int a, bool b) {}
	float g() {}
	int i;
};

struct FakeDer {
	void f(int a, bool b) override{} // no superclass
};

struct Der extends Base {
	void f(int a, int b) override{} // param list doesn't match
	int g() override {} 			// return type doesn't match, (promotable doesn't count)
	void h() override {} 			// no matching name
	int i() override {} 			// base class member not a method
	void f(int a, bool b) {}		// missing 'override' specifier
	float g() override {} 			// OK
	float g() {}					// missing 'override' specifier
};

float g() override {}				// override may only appear in methods