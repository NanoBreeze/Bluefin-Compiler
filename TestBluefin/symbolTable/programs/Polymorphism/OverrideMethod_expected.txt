Override specifier can only be used on methods whose class contains a superclass
Parameter types must match
Return type of methods must match
Base class doesn't contain any symbol with same name
Override of base class member must be a method
All overriding methods must contain 'override' specifier
Override OK!
All overriding methods must contain 'override' specifier
Override specifier may not appear in a function
